File: src/bot/commands/music/spotify/spotify.ts
/**
 * spotify.ts -- The base Spotify command. Gives users easy access
 * to the other commands.
 *
 * Copyright (c) 2019-present Kamran Mackey.
 *
 * Ellie is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ellie is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ellie. If not, see <https://www.gnu.org/licenses/>.
 */

import { Command, Flag } from 'discord-akairo';
import { Message } from 'discord.js';

export default class SpotifyCommand extends Command {
  public constructor() {
    super('spotify', {
      aliases: ['spotify', 'sp', 'splookup'],
      category: 'Music',
      clientPermissions: ['EMBED_LINKS'],
      description: {
        content: 'Gets a variety of information from the Spotify API, such as artist '
          + 'information, album information, song information, and more. ',
        usage: '<subcommand>',
        examples: [
          'artist Alicia Keys',
        ],
      },
      ratelimit: 2,
    });
  }

  public *args() {
    const method = yield {
      type: [
        ['spotify-artist', 'artist'],
      ],
      otherwise: (msg: Message) => {
        const cmdPrefix = this.handler.prefix;
        return msg.channel.send('You did not enter a valid subcommand! Please check '
          + `${cmdPrefix}help spotify to view the available commands.`);
      },
    };

    return Flag.continue(method);
  }
}

File: src/commands/music/spotify/commands/spotify.rs
use serenity::client::Context;
use serenity::framework::standard::macros::command;
use serenity::framework::standard::{CommandError, CommandResult};
use serenity::model::prelude::Message;

use crate::commands::music::spotify::commands::album::*;
use crate::commands::music::spotify::commands::track::*;

use std::env;

#[command]
#[description(
    "Gets a variety of information from the Spotify API, such as \
    artist information, album information, song information, and more."
)]
#[sub_commands(album, track)]
fn spotify(ctx: &mut Context, message: &Message) -> CommandResult {
    let prefix = env::var("DISCORD_PREFIX").unwrap();

    message
        .channel_id
        .send_message(&ctx, move |m| {
            m.embed(move |e| {
                e.title("Error: Invalid / No Subcommand Entered!");
                e.description(format!(
                    "You did not enter a valid subcommand! Please check \
                    `{}help spotify` for the command usage.",
                    prefix
                ));
                e
            })
        })
        .map_or_else(|e| Err(CommandError(e.to_string())), |_| Ok(()))
}
